include/ignition/rendering/Visual.hh

```diff
@@ -28,7 +28,15 @@ namespace ignition
   {
     inline namespace IGNITION_RENDERING_VERSION_NAMESPACE {
     //
-    using Variant = std::variant<int, float, double, std::string>;
+    /// \brief Alias for a variant that can holds various types of data.
+    /// The first type of the variant is std::monostate in order to prevent
+    /// default-constructed variants from holding a type (a default-constructed
+    /// variant is returned when a user calls Visual::UserData with a key that
+    /// doesn't exist for the visual. In this case, since the key doesn't
+    /// exist, the variant that is returned shouldn't hold any types - an
+    /// "empty variant" should be returned for keys that don't exist)
+    using Variant = std::variant<std::monostate, int, float, double,
+          std::string>;
```

Comment:
We can't change this without breaking ABI :confused: 

https://build.osrfoundation.org/job/ignition_rendering-abichecker-any_to_any-ubuntu_auto-amd64/646/API_5fABI_20report/

Reply:
Ohhh you're right! I didn't even think about ABI here. I think that means we can't fix this issue in any released version then :disappointed_relieved: I wish we could at least add a method to the `Visual` class like `HasUserData`, and then encourage users to call this first before calling `UserData` in order to prevent using keys that don't exist. However, doing this would break ABI too, according to the [KDE ABI guide](https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B):
```
You cannot...
* add new virtual functions to non-leaf classes as this will break subclasses
```

So, I don't think we can do anything about this in released versions. Should I re-target this to Fortress? If so, is there anything we should do about this for released versions (update documentation, create an issue, etc.)?

Reply:
Yeah I think we can target this at Fortress. For earlier versions, the only thing I can think of is creating our own "invalid variant" and documenting it. For example, we could return an `std::string("_invalid_")` or something like that. And document it...

Reply:
> we could return a `std::string("_invalid_")` or something like that. And document it

I think this is a good idea, but my only concern is that this is technically a behavioral breaking change, which we try to stay away from for released versions. So, are we even able to do anything about this issue for released versions? Or can it only be addressed starting in Fortress?

Reply:
That's a good point, downstream users may be already compensating for the current behaviour, so we could break them. I'm ok targeting this at Fortress and documenting the migration.

Reply:
Sounds good, I will close this since I have opened a new PR for fortress. See #371